_posts/2024-04-24-highlights-otp-27.md

```diff
@@ -0,0 +1,1048 @@
+printed as
+[actual paper books](https://erlangforums.com/t/old-printed-otp-documentation-cover/1989/2?u=bjorng).
+The last time the books were printed were for the Erlang/OTP R7 released in 2000.
+
+As an example, here is the XML code for
+[`lists:duplicate/2`](https://www.erlang.org/docs/26/man/lists#duplicate-2)
+from Erlang/OTP 26:
+
+```xml
+    <func>
+      <name name="duplicate" arity="2" since=""/>
+      <fsummary>Make <c>N</c> copies of element.</fsummary>
+      <desc>
+        <p>Returns a list containing <c><anno>N</anno></c> copies of term
+          <c><anno>Elem</anno></c>.</p>
+        <p><em>Example:</em></p>
+        <pre>
+> <input>lists:duplicate(5, xx).</input>
+[xx,xx,xx,xx,xx]</pre>
+      </desc>
+    </func>
+```
+
+The XML code was stored in separate files, not in the source
+code. When building the documentation, the function specs from the
+source code would be combined with the text from the documentation
+file. It was the responsibility of the writer to ensure that variables
+mentioned in the documentation body matched the names in the function
+spec.
+
+One thing never said about Erl_Docgen and the old documentation system
+was that it made writing documentation enjoyable and effortless. That
+was one thing we wanted to change with the new documentation system.
+We wanted to make it fun to write documentation, or at least to
+require less attention to tedious details such as using XML tags
+correctly.
+
+In Erlang/OTP 27, the documentation is written in
+[Markdown](https://en.wikipedia.org/wiki/Markdown) and is placed in
+the source code before the function spec and implementation. Here is
+the documentation and implementation of
+[`lists:duplicate/2`](https://www.erlang.org/doc/man/lists#duplicate-2)
+in Erlang/OTP 27:
+
+    -doc """
+    Returns a list containing `N` copies of term `Elem`.
+
+    _Example:_
+
+    ```erlang
+    > lists:duplicate(5, xx).
+    [xx,xx,xx,xx,xx]
+    ```
+    """.
+
+    -spec duplicate(N, Elem) -> List when
+          N :: non_neg_integer(),
+          Elem :: T,
+          List :: [T],
+          T :: term().
+
+    duplicate(N, X) when is_integer(N), N >= 0 -> duplicate(N, X, []).
+
+    duplicate(0, _, L) -> L;
+    duplicate(N, X, L) -> duplicate(N-1, X, [X|L]).
+    ```
+
+The documentation is placed in a
+[triple-quoted string](#triple-quoted-strings)
+following
+the [`-doc` attribute](https://www.erlang.org/eeps/eep-0059).
+
+Having the documentation near the spec makes its easy to ensure that
+the text refers to variables defined in the function spec.
+
+Another goal we had was to replace Erl_Docgen with a tool more widely
+used so that we wouldn't have to carry the entire burden for
+maintaining it. We did that by using
+[ExDoc](https://hexdocs.pm/ex_doc/readme.html), which is also used by
+the [Elixir](https://elixir-lang.org) language and most, if not all,
+Elixir projects.
+
+An issue that arose is whether it's advisable to include user
```

Comment:
do we use the contraction `it's` or `it is`?

Reply:
Yes, we do in blog posts but not in our documentation. Blog posts are a little less formal.